### PR TITLE
Add task to import data from data.gouv.fr

### LINKED ISSUE #150: Let validator harvest datasets, given a list of slugs

### PR DESCRIPTION
**Pull request by l-vincent-l**
Import data from data.gouv.fr

Fixes #150

### CODE REVIEW

File: lib/transport/import_data_service.ex [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+  ## Examples
+
+      iex> ImportDataService.check_download_uri(%{"download_uri" => nil})
+      ["no_download_uri"]
+
+      iex> ImportDataService.check_download_uri(%{"download_uri" => "http"})
+      []
+
+  """
+  def check_download_uri(%{"download_uri" => nil}), do: ["no_download_uri"]
```
Comment (maukoquiroga) [inline lib/transport/import_data_service.ex]: Why a list?

File: lib/transport/import_data_service.ex [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+  ## Examples
+
+      iex> ImportDataService.check_license(%{"license" => "bliblablou"})
+      ["bad_license"]
+
+      iex> ImportDataService.check_license(%{"license" => "odc-odbl"})
+      []
+
+  """
+  def check_license(%{"license" => "odc-odbl"}), do: []
```
Comment (maukoquiroga) [inline lib/transport/import_data_service.ex]: Why a list?

File: lib/transport/import_data_service.ex [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+      {:ok, "https"}
+
+  """
+  def get_url_from_csv(body) do
+    _separators = [?;, ?,]
+    |> Enum.map(&(get_url_from_csv(&1, body)))
+    |> Enum.filter(&(&1 != nil))
+    |> case do
+      [url] -> {:ok, url}
+      [url | _] -> {:ok, url}
```
Comment (maukoquiroga) [inline lib/transport/import_data_service.ex]: This one matches ```[url]``` as well, as you can always define a list by ```[h|t]```, where ```t = []```.

File: lib/transport/import_data_service.ex [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+      ...> |> ImportDataService.get_url_from_csv()
+      {:error, "no column file"}
+
+      iex> "Donnees;format;Download\\r\\nHoraires des lignes TER;GTFS;https\\r\\n"
+      ...> |> ImportDataService.get_url_from_csv()
+      {:ok, "https"}
+
+  """
+  def get_url_from_csv(body) do
+    _separators = [?;, ?,]
```
Comment (maukoquiroga) [inline lib/transport/import_data_service.ex]: No need to name this one, or if you still want to do so, add it as a @module value or something like that.

File: lib/transport/import_data_service.ex [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+      iex> ["name,file\\ntoulouse,http", "stop,lon,lat\\n1,48.8,2.3"]
+      ...> |> ImportDataService.get_url_from_csv()
+      "http"
+
+      iex> 
+      ...> |> ImportDataService.get_url_from_csv()
+      {:error, "no column file"}
+
+  """
+  def get_url_from_csv(bodys) when is_list(bodys) do
```
Comment (maukoquiroga) [inline lib/transport/import_data_service.ex]: bodys => bodies

File: lib/transport/import_data_service.ex [hunk 6]
```diff
@@ -0,0 +1,331 @@
...
+      iex> {:ok, %{headers: [{"Content-Type", "application/zip"}]}}
+      ...> |> ImportDataService.filter_header_csv
+      false
+
+      iex> {:error, "pouet"}
+      ...> |> ImportDataService.filter_header_csv
+      false
+
+  """
+  def filter_header_csv({:ok, %{headers: headers}}) do
```
Comment (maukoquiroga) [inline lib/transport/import_data_service.ex]: Why are you passing on ```{:ok, %{headers: headers}}``` to the function? Filter it out first.

Comment (maukoquiroga) [reply lib/transport/import_data_service.ex]: You're not actually filtering, so function name should rather be something like ```csv?```

File: lib/transport/import_data_service.ex [hunk 7]
```diff
@@ -0,0 +1,331 @@
...
+     Enum.any?(headers,
```
Comment (maukoquiroga) [inline lib/transport/import_data_service.ex]:
```elixir
Enum.any?(headers, fn {k, v} ->
   k == "Content-Type" && String.contains?(v, "csv") 
end)
```

File: lib/transport/import_data_service.ex [hunk 8]
```diff
@@ -0,0 +1,331 @@
...
+     |> Map.put("task_id", Map.get(dataset, "task_id"))
+     |> Map.put("download_uri", get_download_uri(dataset))
+    }
+  end
+
+  def get_cached_dataset(_v), do: {:error, "Dataset needs to be a map"}
+
+  def get_download_uri(%{"resources" => resources}) do
+    cond do
+      (l = get_url(resources, &filter_gtfs/1)) != nil -> l
```
[the remark this answers is not in the document]
Comment (l-vincent-l) [reply lib/transport/import_data_service.ex]: How would your return l without that ?

File: lib/transport/import_data_service.ex [hunk 9]
```diff
@@ -0,0 +1,331 @@
...
+      ...> |> ImportDataService.filter_zip
+      []
+
+      iex> [%{"mime" => "application/zip"}, %{"mime" => "application/neptune"}]
+      ...> |> ImportDataService.filter_zip
+      [%{"mime" => "application/zip"}]
+
+  """
+  def filter_zip(r) do
+    Enum.filter(r, fn %{"mime" => mime} -> mime == "application/zip"
```
[the remark this answers is not in the document]
Comment (l-vincent-l) [reply lib/transport/import_data_service.ex]: It won't match a map with another key

File: lib/transport/import_data_service.ex [hunk 10]
```diff
@@ -0,0 +1,331 @@
...
+      iex> [%{"mime" => "text/cv"}]
+      ...> |> ImportDataService.filter_csv()
+      []
+
+      iex> [%{"mime" => "text/csv"}, %{"mime" => "application/neptune"}]
+      ...> |> ImportDataService.filter_csv()
+      [%{"mime" => "text/csv"}]
+
+  """
+  def filter_csv(r) do
```
[the remark this answers is not in the document]
Comment (l-vincent-l) [reply lib/transport/import_data_service.ex]: It won't match a map with another key

File: lib/transport/import_data_service.ex (hunk 9 above)
[the remark this answers is not in the document]
Comment (maukoquiroga) [reply lib/transport/import_data_service.ex]: OK

File: lib/transport/import_data_service.ex (hunk 10 above)
[the remark this answers is not in the document]
Comment (maukoquiroga) [reply lib/transport/import_data_service.ex]: OK

File: lib/transport/import_data_service.ex (hunk 8 above)
[the remark this answers is not in the document]
Comment (maukoquiroga) [reply lib/transport/import_data_service.ex]: I'd get those values before ```cond```.

**Pull request merged**